src/Builder/AbstractFormContractor.php

```diff
@@ -0,0 +1,197 @@
+        $formOptions = \func_get_args()[2] ?? [];
+
+        $options = [];
+        $options['sonata_field_description'] = $fieldDescription;
+
+        if ($this->isAnyInstanceOf($type, [
+            ModelType::class,
+            ModelTypeList::class,
+            ModelListType::class,
+            ModelHiddenType::class,
+            ModelAutocompleteType::class,
+            ModelReferenceType::class,
+        ])) {
+            // NEXT_MAJOR: Remove this check.
+            if ('list' === $fieldDescription->getOption('edit')) {
+                throw new \LogicException(sprintf(
+                    'The `%s` type does not accept an `edit` option anymore,'
+                    .' please review the UPGRADE-2.1.md file from the SonataAdminBundle',
+                    ModelType::class
+                ));
+            }
+
+            $options['class'] = $fieldDescription->getTargetModel();
+            $options['model_manager'] = $fieldDescription->getAdmin()->getModelManager();
+
+            if ($this->isAnyInstanceOf($type, [ModelAutocompleteType::class])) {
+                if (!$fieldDescription->getAssociationAdmin()) {
+                    throw new \RuntimeException(sprintf(
+                        'The current field `%s` is not linked to an admin.'
+                        .' Please create one for the target model: `%s`',
+                        $fieldDescription->getName(),
+                        $fieldDescription->getTargetModel()
+                    ));
+                }
+            }
+        } elseif ($this->isAnyInstanceOf($type, [AdminType::class])) {
+            if (!$fieldDescription->getAssociationAdmin()) {
+                throw new \RuntimeException(sprintf(
+                    'The current field `%s` is not linked to an admin.'
+                    .' Please create one for the target model : `%s`',
+                    $fieldDescription->getName(),
+                    $fieldDescription->getTargetModel()
+                ));
+            }
+
+            // NEXT_MAJOR: Change this check with: "if (!$fieldDescription->hasSingleValueAssociation())".
+            if (!$this->hasSingleValueAssociation($fieldDescription)) {
+                throw new \RuntimeException(sprintf(
+                    'You are trying to add `%s` field `%s` which is not One-To-One or  Many-To-One.'
+                    .' Maybe you want `%s` instead?',
+                    AdminType::class,
+                    $fieldDescription->getName(),
+                    CollectionType::class
+                ));
+            }
+
+            // set sensitive default value to have a component working fine out of the box
+            $options['btn_add'] = false;
+            $options['delete'] = false;
+
+            $options['data_class'] = $fieldDescription->getAssociationAdmin()->getClass();
+            $options['empty_data'] = static function () use ($fieldDescription) {
```

Comment:
```suggestion
            $options['empty_data'] = static function () use ($fieldDescription): object {
```

Reply:
Like the exception, I've added a comment for the next major version.